src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+      dispatcher_, seastar::make_ready_future<>());
+    stop_ = true;
+    cond_.signal();
+    co_await std::move(dispatcher);
+    co_await gate_.close();
+}
+
+const std::filesystem::path& io_queue::path() const noexcept { return path_; }
+
+seastar::future<> io_queue::open() noexcept {
+    log.debug("Opening {}", path_);
+    assert(!opened());
+    assert(running_.empty());
+    file_ = co_await storage_->open(path_);
+    cond_.signal();
+}
+
+bool io_queue::opened() const noexcept { return file_ != nullptr; }
+
+/*
+ * closing the file always succeeds. once seastar::file::close is invoked the
+ * handle is not longer usable, so we swallow any errors here and release units
+ * to avoid blocking other queues as there isn't much else we can do.
+ *
+ * failure to close a file is quite rare outside of the normal cases like trying
+ * to close an unknown or already closed file handle. these cases won't occur
+ * using the seastar api since it prevents those mistakes. in any case, what we
+ * care about most when closing a file is that its file descriptor resource is
+ * freed in the kernel to avoid reaching an open file limit.
+ */
+seastar::future<> io_queue::close() noexcept {
+    log.debug("Closing {}", path_);
+    assert(opened());
+
+    // ensure all on-going ops have completed
+    auto units = co_await seastar::get_units(
+      ops_, seastar::semaphore::max_counter());
+
+    auto file = std::exchange(file_, nullptr);
+    try {
+        co_await file->close();
+    } catch (...) {
+        log.warn(
+          "Unable to close file {}: {}", path_, std::current_exception());
+    }
+}
+
+void io_queue::submit_read(page& page) noexcept {
+    page.set_flag(page::flags::read);
+    enqueue_pending(page);
+}
+
+void io_queue::submit_write(page& page) noexcept {
+    /*
+     * if the page is not on any list, then in queue it.
+     */
+    if (!page.io_queue_hook.is_linked()) {
+        page.set_flag(page::flags::write);
+        enqueue_pending(page);
+        return;
+    }
+
+    assert(page.test_flag(page::flags::write));
+
+    /*
+     * if it isn't queued, then that means the dispatcher has handled it. in
+     * this case, new data may have arrived into the page. we mark it as queued
+     * as a signal to the I/O completion continuation to requeue the write.
+     */
+    if (!page.test_flag(page::flags::queued)) {
+        log.debug("Marking inflight I/O to be requeued");
+        page.set_flag(page::flags::queued);
+    }
+}
+
+void io_queue::enqueue_pending(page& page) noexcept {
+    pending_.push_back(page);
+    page.set_flag(page::flags::queued);
+    cond_.signal();
+}
+
+seastar::future<> io_queue::dispatch() noexcept {
+    while (true) {
+        co_await cond_.wait([this] { return !pending_.empty() || stop_; });
+
+        if (stop_) {
+            co_return;
+        }
+
+        auto units = co_await seastar::get_units(ops_, 1);
+
+        if (file_ == nullptr) {
+            continue;
+        }
```

Comment:
Question: is it possible to ever end up here, as we open the file in `open()` before signalling `cond_`. 

If this code path is valid, should this be a part of `cond_.wait`?

Reply:
It's definitely possible, the tests crash without it. The intuition behind the need to check this is that `close()`, which clears the file_ pointer, is allowed to run concurrency with the dispatch loop. Since the dispatch loop and `close()` synchronize on the `ops_` semaphore, checking here handles the race where the dispatch loop wakes up after `close()` completed. That's why it can't be part of the cond_.wait predicate only.

Reply:
added a comment into the code

```diff
diff --git a/src/v/io/io_queue.cc b/src/v/io/io_queue.cc
index 6bf1bc477a..64694ceb1f 100644
--- a/src/v/io/io_queue.cc
+++ b/src/v/io/io_queue.cc
@@ -124,6 +124,11 @@ seastar::future<> io_queue::dispatch() noexcept {

         auto units = co_await seastar::get_units(ops_, 1);

+        /*
+         * dispatch loop and close() run concurrently, and synchronize ont he
+         * ops_ semaphore. if dispatch wakes up after close() completed, then we
+         * need to identify that case here--we no longer have a file descriptor.
+         */
         if (file_ == nullptr) {
             continue;
         }
```